src/box/xrow.c

```diff
@@ -1921,26 +1921,97 @@ xrow_encode_join(struct xrow_header *row, const struct tt_uuid *instance_uuid)
 }
 
 int
-xrow_encode_vclock(struct xrow_header *row, const struct vclock *vclock)
+xrow_encode_heartbeat(struct xrow_header *row, const struct vclock *vclock,
+		      uint64_t vclock_sync)
 {
 	memset(row, 0, sizeof(*row));
-
-	/* Add vclock to response body */
-	size_t size = 8 + mp_sizeof_vclock_ignore0(vclock);
+	row->type = IPROTO_OK;
+	size_t size = 0;
+	size_t map_size = 0;
+	if (vclock != NULL) {
+		map_size += 1;
+		size += mp_sizeof_uint(IPROTO_VCLOCK);
+		size += mp_sizeof_vclock_ignore0(vclock);
+	}
+	if (vclock_sync != 0) {
+		map_size += 1;
+		size += mp_sizeof_uint(IPROTO_VCLOCK_SYNC);
+		size += mp_sizeof_uint(vclock_sync);
+	}
+	if (map_size == 0)
+		return 0;
+	size += mp_sizeof_map(map_size);
 	char *buf = (char *) region_alloc(&fiber()->gc, size);
 	if (buf == NULL) {
-		diag_set(OutOfMemory, size, "region_alloc", "buf");
+		diag_set(OutOfMemory, size, "region_alloc",
+			 "heartbeat message buf");
 		return -1;
 	}
 	char *data = buf;
-	data = mp_encode_map(data, 1);
-	data = mp_encode_uint(data, IPROTO_VCLOCK);
-	data = mp_encode_vclock_ignore0(data, vclock);
+	data = mp_encode_map(data, map_size);
+	if (vclock != NULL) {
+		data = mp_encode_uint(data, IPROTO_VCLOCK);
+		data = mp_encode_vclock_ignore0(data, vclock);
+	}
+	if (vclock_sync != 0) {
+		data = mp_encode_uint(data, IPROTO_VCLOCK_SYNC);
+		data = mp_encode_uint(data, vclock_sync);
+	}
 	assert(data <= buf + size);
 	row->body[0].iov_base = buf;
 	row->body[0].iov_len = (data - buf);
 	row->bodycnt = 1;
-	row->type = IPROTO_OK;
+	return 0;
+}
+
+int
+xrow_decode_heartbeat(const struct xrow_header *row, struct vclock *vclock,
```

Comment:
> But why do you want to separate them anyway?

Currently we rely on heartbeats being the same request in both directions. It is not so even now - relay doesn't send vclocks, sadly. In scope of #7253 I will add term to acks, only in one direction too. As a result we would have 3 optional fields: `vclock_sync`, `vclock`, `term`. Their subset would define what request we want to send - applier->relay ack or relay->applier ack. The result will be a mess just like we have at the moment with `xrow_decode_subscribe()` - it is used for almost everything with various subsets of fields, which is super confusing.

What is worse, we keep growing this confusion - for example, `xrow_decode_heartbeat()` is called by `xrow_decode_vclock()` - these are different requests.

If we update heartbeats with a new field `term`, it will affect all their encode/decode calls (even in relay->applier direction) and even non-related code like `xrow_decode_vclock()`.

This is why I would rather split these acks into separate requests/functions, even if there would be some small code duplication like `vclock_sync` which can be sent by either of them optionally.

> Now the protocol is ready for performing the requests in correct direction

Not really, we will still need to change it again. At least, to send vclocks from relay (on demand, not always, right?). However my ranting is rather about the code, not about the protocol. How `xrow_decode_subscribe()` is used is truly terrifying. I would want us to avoid the same for acks/heartbeats.

> But it might need that later

Yes, exactly. So what I want us to try is to isolate such changes as much as possible. If we change messages relay->applier only, then it shouldn't affect the code, which encodes/decodes applier->relay messages.

Anyway, you don't need to do it all now, as I said. I am doing this big rework of xrow.h in my own patchset already. I couldn't stand to add another optional parameter `term` to `xrow_decode_subscribe()`.

Reply:
> In scope of https://github.com/tarantool/tarantool/issues/7253 I will add term to acks, only in one direction too. As a result we would have 3 optional fields: vclock_sync, vclock, term. Their subset would define what request we want to send - applier->relay ack or relay->applier ack. The result will be a mess just like we have at the moment with xrow_decode_subscribe() - it is used for almost everything with various subsets of fields, which is super confusing.

I see now, thanks for the explanation!

>> Now the protocol is ready for performing the requests in correct direction
>
> Not really, we will still need to change it again. At least, to send vclocks from relay (on demand, not always, right?)

Yes, I meant that vclock is an optional field, so adding it later won't break the backward compatibility. Old appliers would ignore it, and new ones would use it.

> I couldn't stand to add another optional parameter term to xrow_decode_subscribe().

Yep, it's a mess already.

By the way, I was thinking, maybe we could use some code generation here to reduce the amount of almost identicat functions? I can't imagine how exactly this would be implemented, but I was thinking something like this. It can't be compiled and I'm not sure that C preprocessor has that functionality.
```c
#define DECODE_REPLICATION(func_name, PARAM_LIST)
void
func_name(struct xrow_header *row, PARAM_LIST)
...
    i = mp_decode_map(...);
    for (j = 0; j < i; j++) {
        key = mp_decode_uint(...);
        switch(key) {
#if IPROTO_VCLOCK in PARAM_LIST
        case IPROTO_VCLOCK:
             //decode vclock
             break;
#endif
#if IPROTO_RAFT_TERM in PARAM_LIST
        case IPROTO_RAFT_TERM:
             //decode term
             break;
#endif
       case default:
            //unknown or unneeded key, ignore
        }
   }
```

And do the same for ENCODE_REPLICATION.

And then we would have one mega-macro ENCODE_REPLICATION and DECODE_REPLICATION and a bunch of
functions
ENCODE_REPLICATION(xrow_encode_subscribe, IPROTO_UUID, IPROTO_VCLOCK, ...)

DECODE_REPLICATION(xrow_decode_heartbeat, IPROTO_VCLOCK, IPROTO_VCLOCK_SYNC, ...)

Reply:
I am doing something like that in my patch. But without compile-time optimizations sadly. That won't be possible with macros in a sane way, but could be possible with C++ templates. We could convert xrow.c to C++ if ever decide we want these opts.

Reply:
> I am doing something like that in my patch. But without compile-time optimizations sadly. 

Ok, looking forward to seeing that in your patch!

> We could convert xrow.c to C++ if ever decide we want these opts.

This might be reasonable. But I'm not sure.